File: src/Tests/Behavioral/GenericFuncDecl/GenericFuncDecl.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

public static (T, T) Swap<T>(T a, T b)
    where T : new()
{
    return (b, a);
}

internal static void Main() {
    nint a = 10;
    nint b = 20;
    (a, b) = Swap<nint>(a, b);
    fmt.Printf("After swap: a=%d, b=%d\n"u8, a, b);
    @string x = "hello"u8;
    @string y = "world"u8;
    (x, y) = Swap(x, y);
    fmt.Printf("After swap: x=%s, y=%s\n"u8, x, y);
}

} // end main_package

File: src/Tests/Behavioral/GenericFuncDecl/GenericFuncDecl.go
package main

import "fmt"

// Swap is a generic function that swaps two values of the same type
func Swap[T any](a, b T) (T, T) {
	return b, a
}

func main() {
	// Call the generic function with int type
	a, b := 10, 20
	a, b = Swap[int](a, b)
	fmt.Printf("After swap: a=%d, b=%d\n", a, b)

	// Type inference allows omitting the type argument
	x, y := "hello", "world"
	x, y = Swap(x, y)
	fmt.Printf("After swap: x=%s, y=%s\n", x, y)
}
